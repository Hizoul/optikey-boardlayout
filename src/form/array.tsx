import { getMapToFromProps, IFieldProps, prependPrefix, SharedField, useArray } from "@xpfw/form"
import { observer } from "mobx-react-lite"
import * as React from "react"
import { Button, IconButton } from "@material-ui/core"
import DeleteIcon from "@material-ui/icons/Delete"
import AddIcon from "@material-ui/icons/Add"

const ArrayField: React.FunctionComponent<IFieldProps> = observer((props) => {
  let mapTo = getMapToFromProps(props)
  const arrayHelper = useArray(props.schema, mapTo, props.prefix)
  let hideArrayType = mapTo.indexOf("keyActions") !== -1 && arrayHelper.length < 2
  console.log("ARRAY ", hideArrayType, mapTo, mapTo.indexOf("keyActions") !== -1, arrayHelper.length)
  return (
    <div>
      {arrayHelper.fields.map((field) => {
        return (
        <div className={`flex flex1 center ${hideArrayType ? "" : "objectMargin"}`} key={prependPrefix(field.mapTo, field.prefix)}>
          <SharedField schema={field.schema} mapTo={field.mapTo} prefix={field.prefix} />
          {hideArrayType ? null : <IconButton
            onClick={field.decreaseSize}
            color="secondary"
          >
            <DeleteIcon />
          </IconButton>}
        </div>
      )})}
      {hideArrayType ? null : 
        <Button
          onClick={arrayHelper.increaseSize}
          variant="contained"
          fullWidth
          endIcon={<AddIcon />}
        >
          Add
        </Button>}
    </div >
  )
})

export default ArrayField
