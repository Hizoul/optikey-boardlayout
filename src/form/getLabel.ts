const getLabel = (label: string) => {
  let shortened = label
  const i = label.lastIndexOf("]")
  if (i !== -1) {
    shortened = shortened.substring(i + 2)
  }
  shortened = shortened.replace(/([A-Z])/g, ' $1')
  shortened = shortened.charAt(0).toUpperCase() + shortened.slice(1)
  return shortened
}

export default getLabel