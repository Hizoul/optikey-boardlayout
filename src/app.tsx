import { FormStore } from "@xpfw/form"
import * as React from "react"
import KeyboardDisplay from "./keyboard"
import RightPanel from "./rightPanel"
import exampleXml, { exampleKeyGroupOverrideXml} from "./xml/example"
import xmlParser from "./xml/parse"

FormStore.setValue("Keyboard", xmlParser(exampleKeyGroupOverrideXml).Keyboard)

const App: React.FunctionComponent<any> = () => (
  <>
    <div className="keyboardContainer">
      <KeyboardDisplay />
    </div>
    <RightPanel />
  </>
)

export default App
