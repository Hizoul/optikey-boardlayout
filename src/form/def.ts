import { ComponentRegistry, ExtendedJSONSchema } from "@xpfw/form"
import { registerComponents } from "@xpfw/form-web"
import BooleanField from "./boolean"
import SelectField from "./select"
import TextField from "./text"
import ObjectField from "./object"
import ArrayField from "./array"

registerComponents()
ComponentRegistry.registerComponent("number", TextField)
ComponentRegistry.registerComponent("string", TextField)
ComponentRegistry.registerComponent("boolean", BooleanField)
ComponentRegistry.registerComponent("object", ObjectField)
ComponentRegistry.registerComponent("array", ArrayField)
ComponentRegistry.registerComponent("string", SelectField, "select")
const actionNames = ["Action", "ChangeKeyboard", "KeyDown", "KeyUp", "Loop", "Text", "Wait"]
const keyActionTypes: any[] = actionNames.map((v) =>{return {value: v, label: v}})
const keyActionObjectSchema = {
  type: "object",
  title: "keyActionObject",
  properties: {
    actionType: {
      type: "string", theme: "select", selectOptions: keyActionTypes,
      description: "The action type determines what kind of input is sent to the OS on key press."}
  }
}

const nameSchema: ExtendedJSONSchema = {
  title: "Name", type: "string",
  description: "The name that will be shown in OptiKey"
}

const widthSchema: ExtendedJSONSchema = {
  title: "Width", type: "string",
  description: "Percent of screen. if negative then use the sum of the value and the screen's dimension"
}

const heightSchema: ExtendedJSONSchema = {
  title: "Height", type: "string",
  description: "Percent of screen. if negative then use the sum of the value and the screen's dimension"
}
const horizontalSchema: ExtendedJSONSchema = {
  title: "HorizontalOffset", type: "string",
  description: "Percent of screen. if negative then use the sum of the value and the screen's dimension"
}

const verticalSchema: ExtendedJSONSchema = {
  title: "VerticalOffset", type: "string",
  description: "Percent of screen. if negative then use the sum of the value and the screen's dimension"
}

const symbolSchema: ExtendedJSONSchema = {
  title: "SymbolMargin", type: "number",
  description: "Amount of empty space left for symbols"
}

const borderSchema: ExtendedJSONSchema = {
  title: "BorderThickness", type: "string",
  description: "Controls how thick the lines between individual keys are"
}

const rowsSchema: ExtendedJSONSchema = {
  title: "Grid.Rows", type: "number",
  description: "Controls the amount of keyboard rows"
}

const colsSchema: ExtendedJSONSchema = {
  title: "Grid.Cols", type: "number",
  description: "Controls the amount of keyboard columns"
}

const persistNewStateSchema: ExtendedJSONSchema = {
  title: "PersistNewState", type: "boolean",
  description: "Indiates whether the size and position of this keyboard will be saved"
}

const windowStateSchema: ExtendedJSONSchema = {
  title: "WindowState", type: "string", theme: "select",
  selectOptions: [
    {label: "Floating", value: "Floating"},
    {label: "Docked", value: "Docked"},
    {label: "Maximised", value: "Maximised"}],
  description: "Controls how the OptiKey window should behave"
}
const hideFromListSchema: ExtendedJSONSchema = {
  title: "HideFromKeyboardMenu", type: "boolean",
  description: "controls whether or not this keyboard will be listed in the dynamic keyboard selection menu. You may wish to set this equal to False for smaller utility-type keyboards."
}

const dockSizeSchema: ExtendedJSONSchema = {
  title: "DockSize", type: "string", theme: "select",
  selectOptions: [
    {label: "Full", value: "Full"},
    {label: "Collapsed", value: "Collapsed"}
  ],
  description: "Controls the size of the dock"
}

const positionSchema: ExtendedJSONSchema = {
  title: "Position", type: "string", theme: "select",
  description: "Controls where the Keyboard will be shown",
  selectOptions: [
    {label: "TopLeft", value: "TopLeft"},
    {label: "Top", value: "Top"},
    {label: "TopRight", value: "TopRight"},
    {label: "Left", value: "Left"},
    {label: "Right", value: "Right"},
    {label: "BottomLeft", value: "BottomLeft"},
    {label: "Bottom", value: "Bottom"},
    {label: "BottomRight", value: "BottomRight"}
  ],
}

const backgroundColorSchema: ExtendedJSONSchema = {
  title: "BackgroundColor", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const borderColorSchema: ExtendedJSONSchema = {
  title: "BorderColor", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const keyboardPrefix = "Keyboard"
const hoverKey = `hover`
const dragKey = `drag`
const activeKey = `active`

export {
  nameSchema, heightSchema, borderSchema, symbolSchema, rowsSchema, colsSchema, keyboardPrefix,
  activeKey, dragKey, hoverKey, persistNewStateSchema, windowStateSchema,
  dockSizeSchema, positionSchema, widthSchema, horizontalSchema, verticalSchema,
  backgroundColorSchema, borderColorSchema, hideFromListSchema, keyActionObjectSchema, actionNames
}
