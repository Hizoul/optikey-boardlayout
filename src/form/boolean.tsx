import Checkbox from '@material-ui/core/Checkbox'
import FormControl from '@material-ui/core/FormControl'
import FormHelperText from '@material-ui/core/FormHelperText'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import { getMapToFromProps, IFieldProps, useFieldWithValidation } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"
import * as React from "react"
import getLabel from "./getLabel"
import { InputLabel } from '@material-ui/core'

const changeWrapper = (fieldHelper: {setValue: Function, value: any}) => {
  return (e: any) => {
    fieldHelper.setValue(!fieldHelper.value)
  }
}

const BooleanField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useFieldWithValidation(props.schema, mapTo, props.prefix)
  const setBool = changeWrapper(fieldHelper)
  return (
    <div className="flex horizontal inputMargin">
      <FormControlLabel
          control={
            <>
            <Checkbox
              checked={fieldHelper.value}
              onChange={setBool}
              value={fieldHelper.value}
            />
            </>
          }
          label={getLabel(mapTo)}
      />
      <FormHelperText className="center">{get(props, "schema.description")}</FormHelperText>
    </div>
  )
})

export default BooleanField
export {
  changeWrapper
}