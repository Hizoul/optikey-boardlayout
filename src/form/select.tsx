import FormControl from '@material-ui/core/FormControl'
import FormHelperText from '@material-ui/core/FormHelperText'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import CheckBoxIcon from '@material-ui/icons/CheckBox'
import { getMapToFromProps, IFieldProps, JSONSchemaDefinition, memo, useFieldWithValidation, prependPrefix, FormStore } from "@xpfw/form"
import { get, isFunction } from "lodash"
import { observer } from "mobx-react"
import * as React from "react"
import getLabel from "./getLabel";
import { toJS } from "@xpfw/data"

const changeWrapperSelect = (fieldHelper: any, mapToOrig: string, prefix?: string) => {
  const actualMapTo = mapToOrig.substring(mapToOrig.lastIndexOf(".") + 1)
  const mapTo = mapToOrig.substring(0, mapToOrig.lastIndexOf("."))
  return (e: any) => {
    const t: any = get(e, "nativeEvent.target")
    const value = t.getAttribute("data-value")
    if (value == null || value == "") {
      let currentVal = FormStore.getValue(mapTo, prefix)
      if (currentVal != null) {
        currentVal = toJS(currentVal)
        delete currentVal[actualMapTo]
        fieldHelper.setValue("")
        FormStore.setValue(mapTo, currentVal, prefix)
      }
    } else {
      fieldHelper.setValue(value)
    }
  }
}

const SelectField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useFieldWithValidation(props.schema, mapTo, props.prefix)
  let changeFunc = changeWrapperSelect(fieldHelper, mapTo, props.prefix)
  let selOpts = get(props, "schema.selectOptions", [])
  if (isFunction(selOpts)) {
    selOpts = selOpts(fieldHelper.value, props.schema, props)
  }
  const options = selOpts.map((option: any) => {
    return (
      <MenuItem key={option.value} value={option.value}>
        {option.label}
      </MenuItem>
    )
  })
  return (
    <FormControl fullWidth className="inputMargin">
      <InputLabel>{getLabel(mapTo)}</InputLabel>
        <Select
          onChange={changeFunc}
          value={fieldHelper.value}
        >
          {options}
        </Select>
        <FormHelperText>{get(props, "schema.description")}</FormHelperText>
     </FormControl>
  )
})

export default SelectField
export {
  changeWrapperSelect
}