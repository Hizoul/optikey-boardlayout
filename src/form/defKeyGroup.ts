import { ExtendedJSONSchema } from "@xpfw/form"
import { backgroundColorSchema } from "./def"

const keyDisabledBackgroundSchema: ExtendedJSONSchema = {
  title: "KeyDisabledBackground", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const keyDisabledForegroundSchema: ExtendedJSONSchema = {
  title: "KeyDisabledForeground", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const keyDisabledOpacitySchema: ExtendedJSONSchema = {
  title: "KeyDisabledOpacity", type: "number"
}

const keyDownBackgroundSchema: ExtendedJSONSchema = {
  title: "KeyDownBackground", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const keyDownForegroundSchema: ExtendedJSONSchema = {
  title: "KeyDownForeground", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const keyDownOpacitySchema: ExtendedJSONSchema = {
  title: "KeyDownOpacity", type: "number"
}

const foregroundColorSchema: ExtendedJSONSchema = {
  title: "ForegroundColor", type: "string", theme: "color",
  description: "for valid colors refer to https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color"
}

const opacitySchema: ExtendedJSONSchema = {
  title: "Opacity", type: "number"
}

const autoScaleWidthSchema: ExtendedJSONSchema = {
  title: "AutoScaleToOneKeyWidth", type: "boolean",
  description: "only required to modify scaling behavior when key width is greater than 1. if not declared it will default to true"
}

const autoScaleHeightSchema: ExtendedJSONSchema = {
  title: "AutoScaleToOneKeyHeight", type: "boolean",
  description: "only required to modify scaling behavior when key width is greater than 1. if not declared it will default to true"
}
const sharedSizeSchema: ExtendedJSONSchema = {
  title: "SharedSizeGroup", type: "boolean",
  description: "only required to break out a key, or set of keys, to size separately, otherwise size grouping is determined automatically"
}

const usePersianSchema: ExtendedJSONSchema = {
  title: "UsePersianCompatibilityFont", type: "boolean"
}

const useUnicodeSchema: ExtendedJSONSchema = {
  title: "UseUnicodeCompatibilityFont", type: "boolean"
}

const useUrduSchema: ExtendedJSONSchema = {
  title: "UseUrduCompatibilityFont", type: "boolean"
}

const lockOnSchema: ExtendedJSONSchema = {
  title: "LockOnTime", type: "number",
  description: "milliseconds from when gaze is detected on key to the start of the progress animation"
}

const completionTimeSchema: ExtendedJSONSchema = {
  title: "CompletionTime", type: "number",
  description: "milliseconds from the start of the progress animation to the key being triggered"
}
const repeatDelaySchema: ExtendedJSONSchema = {
  title: "RepeatDelay", type: "number",
  description: "milleseconds between the initial keystroke and a second consecutive keystroke of the same key"
}
const repeatRateSchema: ExtendedJSONSchema = {
  title: "RepeatRate", type: "number",
  description: "milliseconds between additional consecutive keystrokes of the same keymilliseconds from when gaze is detected on key to the start of the progress animation"
}


const keyGroupSchema: ExtendedJSONSchema = {
  type: "object",
  title: "KeyGroup",
  properties: {
    [String(backgroundColorSchema.title)]: backgroundColorSchema,
    [String(foregroundColorSchema.title)]: foregroundColorSchema,
    [String(keyDisabledBackgroundSchema.title)]: keyDisabledBackgroundSchema,
    [String(keyDisabledForegroundSchema.title)]: keyDisabledForegroundSchema,
    [String(keyDisabledOpacitySchema.title)]: keyDisabledOpacitySchema,
    [String(keyDownBackgroundSchema.title)]: keyDownBackgroundSchema,
    [String(keyDownForegroundSchema.title)]: keyDownForegroundSchema,
    [String(keyDownOpacitySchema.title)]: keyDownOpacitySchema,
    [String(opacitySchema.title)]: opacitySchema,
    [String(autoScaleWidthSchema.title)]: autoScaleWidthSchema,
    [String(autoScaleHeightSchema.title)]: autoScaleHeightSchema,
    [String(sharedSizeSchema.title)]: sharedSizeSchema,
    [String(usePersianSchema.title)]: usePersianSchema,
    [String(useUnicodeSchema.title)]: useUnicodeSchema,
    [String(useUrduSchema.title)]: useUrduSchema,
    [String(lockOnSchema.title)]: lockOnSchema,
    [String(completionTimeSchema.title)]: completionTimeSchema,
    [String(repeatDelaySchema.title)]: repeatDelaySchema,
    [String(repeatRateSchema.title)]: repeatRateSchema
  }
}

export {
  keyGroupSchema
}