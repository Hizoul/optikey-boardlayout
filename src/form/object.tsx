import { getMapToFromProps, IFieldProps, SharedField, useObject, FormStore } from "@xpfw/form"
import { observer } from "mobx-react-lite"
import * as React from "react"
import actionKeyList from "../util/actionKeys"
import { keyActionObjectSchema, actionNames} from "./def"

const ObjectField: React.FunctionComponent<IFieldProps> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useObject(props.schema, mapTo, props.prefix)
  let subFieldContent
  if (props.schema != null && props.schema.title != null && props.schema.title.indexOf("keyActionObject") !== -1) {
    const currentSubType = FormStore.getValue(mapTo, props.prefix)
    if (currentSubType != null) {
      if (actionNames.indexOf(currentSubType.actionType) !== -1) {
        let type = "string"
        let theme = "default"
        let selectOptions: any = null
        if (currentSubType.actionType == "Wait") {
          type = "number"
        } else if (currentSubType.actionType == "Action") {
          theme = "select"
          selectOptions = actionKeyList.map((v) => ({label: v, value: v}))
        }
        subFieldContent = <SharedField schema={{type, title: "value", theme, selectOptions}} mapTo={`${mapTo}.value`} prefix={props.prefix} />
        if (currentSubType.actionType == "ChangeKeyboard") {
          subFieldContent = (
            <>
              {subFieldContent}
              <SharedField schema={{type: "boolean", title: "BackReturnsHere"}} mapTo={`${mapTo}.BackReturnsHere`} prefix={props.prefix} />
            </>
          )
        }
        if (currentSubType.actionType == "Loop") {
          subFieldContent = (
            <>
              <SharedField schema={{type: "number", title: "Wait"}} mapTo={`${mapTo}.Wait`} prefix={props.prefix} />
              <SharedField schema={{type: "number", title: "Count"}} mapTo={`${mapTo}.Count`} prefix={props.prefix} />
              <SharedField schema={{type: "array", items: keyActionObjectSchema, title: "keyActions"}} mapTo={`${mapTo}.keyActions`} prefix={props.prefix} />
            </>
          )
        }
      }
    }
  }  
  return (
    <div className="flex1 vertical">
      {fieldHelper.fields.map((subField) => <SharedField {...subField} key={subField.mapTo} />)}
      {subFieldContent}
    </div>
  )
})

export default ObjectField
