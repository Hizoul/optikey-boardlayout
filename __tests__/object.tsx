import * as React from "react"
import { doParsing } from "../src/app"
import renderSnapshot from "../src/testUtil/renderSnapshot"
import ObjectField from "../src/form/object"
import { keyboardPrefix } from "../src/form/def"
import exampleXml from "../src/xml/example"
import { FormStore } from "@xpfw/form/dist"

test("app", () => {
  doParsing(exampleXml)
  const objSchema: any = {
    type: "object",
    title: "keyActionObject",
    properties: {
      bla: {type: "string"},
      blu: {type: "string"}
    }
  }
  renderSnapshot(<ObjectField schema={objSchema} prefix={keyboardPrefix} />, "without actionType specified")
  FormStore.setValue(objSchema.title, {actionType: "unknown"}, keyboardPrefix)
  renderSnapshot(<ObjectField schema={objSchema} prefix={keyboardPrefix} />, "with invalid actionType specified")
})
