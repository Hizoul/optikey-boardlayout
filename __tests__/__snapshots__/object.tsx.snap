// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app: with invalid actionType specified 1`] = `
<div
  className="flex1 vertical"
>
  <div
    className="MuiFormControl-root inputMargin"
  >
    <input
      fullWidth={true}
      inputProps={
        Object {
          "max": undefined,
          "min": undefined,
          "step": undefined,
        }
      }
      label="Key Action Object.bla"
      onChange={[Function]}
      type="text"
      value=""
    />
    <p
      className="MuiFormHelperText-root"
    />
  </div>
  <div
    className="MuiFormControl-root inputMargin"
  >
    <input
      fullWidth={true}
      inputProps={
        Object {
          "max": undefined,
          "min": undefined,
          "step": undefined,
        }
      }
      label="Key Action Object.blu"
      onChange={[Function]}
      type="text"
      value=""
    />
    <p
      className="MuiFormHelperText-root"
    />
  </div>
</div>
`;

exports[`app: without actionType specified 1`] = `
<div
  className="flex1 vertical"
>
  <div
    className="MuiFormControl-root inputMargin"
  >
    <input
      fullWidth={true}
      inputProps={
        Object {
          "max": undefined,
          "min": undefined,
          "step": undefined,
        }
      }
      label="Key Action Object.bla"
      onChange={[Function]}
      type="text"
      value=""
    />
    <p
      className="MuiFormHelperText-root"
    />
  </div>
  <div
    className="MuiFormControl-root inputMargin"
  >
    <input
      fullWidth={true}
      inputProps={
        Object {
          "max": undefined,
          "min": undefined,
          "step": undefined,
        }
      }
      label="Key Action Object.blu"
      onChange={[Function]}
      type="text"
      value=""
    />
    <p
      className="MuiFormHelperText-root"
    />
  </div>
</div>
`;
